---
title:"SLE712 Assessment 3- Part 2-Determine the limits of BLAST"
author:"Yasodha Jayawardena & Dhanusha Weerawardana 
output:html_document
theme: cosmo 
---
GitHub repository: https://github.com/Dhanusha782/SLE712---Bioinformatics-And-Molecular-Biology-Tec.-Assesment-3-part-2ni.git

# Introduction

BLAST is a program or algorithm which is important combinaion of biology and technology to compare biological sequences including, amino acid sequences, and DNA and RNA nucleotied sequences. During this assesment, students able to gain knowladge on BLAST program as wel as able to determine the limits of BLAST. Therefore, in order to compare the sequences, E.Coli gene sequence and Student56 gene sequence have used through out the process. 

## Prerequisite Libraries
There are three packages that we will need to get things working. 
`seqinr` package designed for sequence data procesing and analysis
`R.utils` package designed for extract data from compressed files
`rBLAST` package provide R interface for BLAST search
Additionally, we source two functions which supplied for this practical to perform an analysis into the limits of BLAST.

```{r,chunk1 libraries,echo=TRUE}
#libraries that we need
library("seqinr")
library("R.utils")
library("rBLAST")
```

## Question 1-Download the whole set of E. coli gene DNA sequences and use gunzip to decompress to create a blast database.

Here we are used the `download.file()` command to obtain the E.coli dataset from ensemble FTP page and then we are decompress the file using gunzip command. `makeblast()` function use to creat a BLAST database.
*Number of sequences are present in the E.coli set= 4140
```{r,chunk2 downlod  E.coi gene,echo=TRUE}
#Download the E.coli gene sequency from the Ensemble FTP page
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
```
Use `gunzip()` command to uncompress file and `makeblastdb()` to create blast db
```{r,chunk3 gunzip,echo=TRUE}
#use gunzip to uncompress the file
gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",overwrite=TRUE)
#create a blast database with makeblast function
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa",dbtype = "nucl","-parse_seqids")
```

## Question 2-Download the sample fasta sequences and read into R and determine the length and propostion of GC bases 

Here we are used the `download.file()` command to obtain the sample file and then read it into R. The `str` use to determine data structure. Next select allocated sequence 56 and read it as "myEcoli" and check its structure. Allocated sequence is 56
```{r,chunk4 downlod sample.fa and read,echo=TRUE}
#Download sample file
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa", destfile = "sample.fa")
#Read file into R as Ecoli
Ecoli<-read.fasta("sample.fa")
str(Ecoli) #check data structure
#Read my allocated sequence (56) into R as myEcoli
myEcoli<-Ecoli[[56]]
#Lets check if the data has been imported properly
myEcoli
str(myEcoli)
```
Determine length and propotion of GC bases in allocated sequence by using `getLength()` command and `GC()` command respectively.
*Length=276 bp
*The propotion of GC bases= 0.5326087
```{r,chunk5 length and GC,echo=TRUE}
#calculate length for my allocated gene
seqinr::getLength(myEcoli)
#calculate GC content for my allocated gene
seqinr::GC(myEcoli)
```

## Question 3-Use Blast to identify what E. coli gene your sequence matches best and show a table of percent identity, E-value and bit scores for top3 hits.

In order to creat a BLAST database, run the BLAST by the `download.file(location of the file)` command. Then `res<-` function can be applied to test the command and `head()` and `str` commands can be used to arrenge the table acoording to structure. 

```{r,chunk6 BLAST databese,echo=TRUE}
#sourced the function that ran blast searches in R
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R",destfile = "mutblast.R")
source("mutblast.R")
#test the function
res<-myblastn_tab(myseq = myEcoli,db= "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
#have a look at the blast results
head(res)
str(res)
```
The first top 3 hits can be obtain by creating a character `hits` and applying running command for the hits.
```{r,chunk7 top 3 hits,echo=TRUE}
#identify what E.coil gene matches best and the top 3 hits
hits <-as.character(res$sseqid[1:3])
hits
# show a percent identity, E-value and bit scores of the top 3 hits
tophits <- db[which(names(db) %in% hits[1:3])]
tophits[1:3]
```

The `seqinr` package has to be run and it is a Exploratory data analysis and data visualization for biological sequence and then created a blast database by the command `makeblast`.
Thereafter the `cat` command can be applied to produce the out by converting argument in to a character vectors.
```{r,chunk8 blast db,echo=TRUE}
#create a Blast database
seqinr::write.fasta(tophit,names=names(tophit),file.out = "tophit.fa")
makeblastdb("tophit.fa", dbtype = "nucl","-parse_seqids")
res <- myblastn(myseq = myEcoli, db= "tophit.fa")
res
cat(res,fill=TRUE)
```

##Question 4-Write R code to check the number of mismatches between original and mutated sequence

Give the command `mutator` to extract mutation with 30 substitutions 
```{r,chunk9 mismatches,echo=TRUE}
#create a mutated copy with 30 substitutions
myEcolimutator<- mutator(myEcoli,30)
#chech number of mismatches between the original and mutated sequence
res <- myblastn_tab(myseq = myEcolimutator, db= "tophit.fa")
res
```

##Question 5-Determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin.

```{r,chunk10 mismatches,echo=TRUE}
#test with mismatches
myEcolimutator <- mutator(myEcoli,10)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,20)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,30)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,40)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,50)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,60)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,70)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
myEcolimutator <- mutator(myEcoli,80)
res<-myblastn_tab(myseq = myEcolimutator,db= "myEcoli.fa")
res
```
In this section we have created a function, which is to sumarise and reports risults as 0 or 1 and tested the function by the command `myfunc`. The commanded number or mutation is 50.

```{r,chunk11 mutator,echo=TRUE}
mutator #randomize with mutator
myblastn_tab
#create a function to summerise the result and report as a 0 or 1
myfunc <- function (myseq,nmut){
  mutseq <- mutator(myseq=myseq,nmut=nmut)
  res<-myblastn_tab(myseq=mutseq, db= "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
  if(is.null(res)){myres=0} else {myres=1}
  return(myres)}

#test the function
myfunc(myseq=myEcoli,nmut=50)
#repeat this routine 100 times to grt accurate result
replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=50))
```
In order to acive a reliable answer test must run multiple times, thereby `replicate` command applied to run the test repeatedly and number of repeated test given as 100. Thereafter the results summarized into decimal numbers in between 0 and 1 to get a proportion the number of sucessful BLASTs. Then the interest number of mutant to be evaluated comanded by `n<-c` code and gave function to run multiple tests for all the interst values. Finally, by the command `finalres` tested the BLASTs results.
```{r,chunk12,echo=TRUE}
#summerise result to decimal number(between 0 and 1),to get a propostion of how many BLASTs were successful
mean (replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=50)))
#all the values of nmut that would like to evaluate
n<-c(0,10,20,30,40,50,60,70,80,90,100)
#run replicate command to all those values
myfunc_rep <-function (nmut){mean (replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=50)))}
#test the BLASTs result
finalres <-sapply(n,myfunc_rep)
finalres
```
In this section we have commanded same fuction for the full sequence.
```{r,chunk13,echo=TRUE}
#Follow same method for full sequence
myfunc(myseq=myEcoli,nmut=276)
replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=276))
mean (replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=276)))
n<-c(0,10,20,30,40,50,60,70,80,90,100)
myfunc_rep <-function (nmut){mean (replicate(n=100,expr = myfunc(myseq=myEcoli,nmut=276)))}
finalres <-sapply(n,myfunc_rep)
finalres
```

##Question -Chart or table that shows how the increasing proportion of mutated bases reduces the ability for BLAST to match the gene of origin

```{r,chink14,echo=TRUE}
#create a table to show increasing propotion of mutated bases reduce the ability for BLAST
Bdata <- c(0,10,20,30,40,50,60,70,80,90,100)
Bdata <- as.data.frame(Bdata)
Bdata$repsite<-c(1,1,1,0.95,0.75,0.49,0.25,0.12,0.05,0.02,0.01)
Bdata
##create a graph to show increasing propotion of mutated bases reduce the ability for BLAST
plot(Bdata, type="b", col="red",col.main="blue", col.sub="green",
     main="How increasing no.random bases affects BLAST performance",sub="100 repeates,using sequence number 56",
     xlab="Number of site randomised", ylab="Propotion of successful BLASTs"
     )
```
According to the provide table and graph, when increasing proportion of mutated bases reduce the ability for BLAST to match the gene of origin. When the no.of site are low, performance of BLAST nearly 100%. Half of the queris failing, naerly quator of bases should be need to modify and and when 80-100 sites randomise, the successful rate is very low. 

